kernel/src/drivers/ps2/mod.rs

```diff
@@ -90,167 +93,159 @@ impl Ps2Controller {
-    }
-}
+    pub fn read_config(&mut self) -> Result<(), Ps2Error> {
+        let read = command_ret(ControllerReturnCommand::ReadConfig)?;
+        self.config = Ps2ConfigFlags::from_bits_truncate(read);
 
-/// Returns if the given code is OK
-#[allow(dead_code)] // To be used by keyboard / mouse driver depending on PS/2
-pub fn is_ok(code: Option<u8>) -> bool {
-    code == Some(ACK) || code == Some(SELF_TEST_PASSED)
+        Ok(())
+    }
 }
 
 /// Represents a PS2 device
 pub struct Ps2Device {
-    flags: u8,
+    state: DeviceState,
+    port: DevicePort,
 }
 
 impl Ps2Device {
+    const fn new(port: DevicePort) -> Self {
+        Ps2Device {
+            state: DeviceState::Unavailable,
+            port,
+        }
+    }
+
     /// Tests this device to see if it is available
-    pub fn test(&mut self) -> bool {
-        let available = command_ret(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn test(&mut self) -> Result<bool, Ps2Error> {
+        let available = command_ret(if self.port == DevicePort::Mouse {
             ControllerReturnCommand::TestPort2
         } else {
             ControllerReturnCommand::TestPort1
-        }) == Some(0x0);
-        self.set_flag(DEVICE_AVAILABLE_FLAG, available);
-        available
+        })? == 0x0;
+
+        if available {
+            self.state = DeviceState::Available;
+        } else {
+            self.state = DeviceState::Unavailable;
+        }
+
+        Ok(available)
     }
 
     /// Enables this device
-    pub fn enable(&mut self) {
-        if self.get_flag(DEVICE_AVAILABLE_FLAG) {
-            command(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn enable(&mut self) -> Result<(), Ps2Error> {
+        if self.state == DeviceState::Available {
+            command(if self.port == DevicePort::Mouse {
                 ControllerCommand::EnablePort2
             } else {
                 ControllerCommand::EnablePort1
-            });
-            self.set_flag(DEVICE_ENABLED_FLAG, true);
+            })?;
+            self.state = DeviceState::Enabled;
         }
+        Ok(())
     }
 
     /// Disables this device
-    pub fn disable(&mut self) {
-        command(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn disable(&mut self) -> Result<(), Ps2Error> {
+        command(if self.port == DevicePort::Mouse {
             ControllerCommand::DisablePort2
         } else {
             ControllerCommand::DisablePort1
-        });
-        self.set_flag(DEVICE_ENABLED_FLAG, false);
+        })?;
+        self.state = DeviceState::Available;
+        Ok(())
     }
 
     /// Resets this device
-    pub fn reset(&mut self) {
-        self.command(DeviceCommand::Reset);
+    pub fn reset(&mut self) -> Result<(), Ps2Error> {
+        self.command(DeviceCommand::Reset)?;
+
+        Ok(())
     }
 
     /// Sends a command for this PS2 device and returns result
     #[allow(dead_code)] // Used by drivers depending on PS/2
```

Comment:
Does this need to be allow'ed anymore? We should look over all allows in this module

Reply:
They are allowed because the API is not used yet and that's out of scope.

Reply:
I could've sworn that it is used here though...

Reply:
I don't believe so.